### DISCUSSION TITLE
"auto-session ERROR: Error restoring session, disabling auto save." with "Error: E184: No such user-defined command: ShFoldIfDoFor"

### DISCUSSION THREAD

**Discussion by miguno** [rmagatti/auto-session#358] [Q&A] [answered]
I'm on the currently latest version of auto-session (commit `a5750c6`).

## Expected Behavior

auto-session automatically restores a session, just like it has done for months without any problems.

## Actual Behavior

The following error message is shown when a session is being restored. From what I can tell, the session is still correctly restored despite what the error message says. However, auto save is indeed disabled.

I have not seen any such error message before, and `ShFoldIfDoFor` does not seem to be a part of the auto-session code base either (maybe it's related to folding of code blocks).

```
auto-session ERROR: Error restoring session, disabling auto save.
Set silent_restore = false in the config for a more detailed error message.
Error: E184: No such user-defined command: ShFoldIfDoFor
```

When I set `silent_restore = false` in the auto-session config, then the session is apparently still correctly restored, and (surprise) no error message is shown at all. Also, auto save is now correctly working again in this case. Right now, I am using `silent_restore = false` as a temporary workaround to make auto-session work as expected, but I am confused about why this workaround is even ale to solve the reported issue.

No info in log files: Regardless of the setting of this config option, there's no additional information in neovim's log files about this error as far as I can tell.

Has anyone come across this before?

## Environment

* macOS 14.6.1 (23G93)
* wezterm `20240203-110809-5046fc22` as terminal
* neovim 0.10.0
    ```
    NVIM v0.10.0
    Build type: Release
    LuaJIT 2.1.1713484068
    Run "nvim -V1 -v" for more info
    ```
* lazy.nvim for plugin management
* auto-session at commit a5750c6

auto-session configuration:

```lua
return {
  {
    "rmagatti/auto-session",
    dependencies = {
      "rcarriga/nvim-notify",
      "nvim-telescope/telescope.nvim",
    },
    opts = {
      log_level = "error",
      auto_session_suppress_dirs = {
        "~/Downloads",
        "/",
        "/tmp",
      },
      post_delete_cmds = {
        [[lua vim.notify('Editor session deleted!', vim.log.levels.INFO, { title = 'Neovim' })]],
      },
      post_restore_cmds = {
        [[lua vim.notify('Editor session restored!', vim.log.levels.INFO, { title = 'Neovim' })]],
      },
      post_save_cmds = {
        [[lua vim.notify('Editor session saved!', vim.log.levels.INFO, { title = 'Neovim' })]],
      },
      -- session-lens related settings; nowadays merged into auto-session,
      -- originally at https://github.com/rmagatti/session-lens (including docs)
      session_lens = {
        previewer = false,
        prompt_title = "Editor Sessions",
      },
    },
  },
}

```

**Comment by cameronr** [comment]
Hmm, very interesting. `ShFoldIfDoFor` is part of vim's builtin syntax highlighting for shells scripts so definitely not an `auto-session` thing. Maybe we're inadvertently picking up an error that we shouldn't be. 

Can you post the session file that generates the error when restoring? Should be in `~/.local/share/nvim/sessions/`

**Reply by cameronr** [reply]
This is reproducing for me so I'll dig into it

**Reply by cameronr** [reply, marked as answer]
Ok, I got to the bottom of it. The vim syntax file for sh scripts has these lines that are clearing some commands just in case they exist:

https://github.com/neovim/neovim/blob/6af9ca4926e05ac698d828ab6aeef2f45ccccdf8/runtime/syntax/sh.vim#L109-L113

That doesn't generate an error (because of the `sil!`) but it does set `v:errmsg`. When we're restoring a session, we use `silent! source <session>`. the `!` means it will keep restoring the session even if there's an error so we use `v:errmsg` to catch any errors that might be generated while restoring the session. This case highlights the problem with that approach, namely that there might be other benign "errors" that set `v:errmsg` but aren't really errors. I have an idea on how to not get tripped up by these non-errors. 

In the meantime, setting `silent_restore = false` if you run into this problem is the right approach.

**Reply by miguno** [reply]
Thanks a lot @cameronr for taking your time to investigate this and finding the answer. I really appreciate it!

Like you suggested, I'll then keep using the workaround of setting `silent_restore = false` until the root cause is fixed in auto-session.

**Reply by cameronr** [reply]
I've pushed my changes so you can remove setting `silent_restore` from your config.

As part of the changes, a bunch of the config values were renamed/changed so `silent_restore` is now `continue_restore_on_error` but leaving it at it's default (of `true`) should no longer throw an error when loading an shell script.

**Reply by miguno** [reply]
Thanks for the update, @cameronr! The latest version of auto-session works for me now, and the workaround is indeed no longer needed.

**Reply by savchenko** [reply]
Was the `Error: E184: No such user-defined command: ShFoldIfDoFor` ever reported upstream?

**Reply by cameronr** [reply]
I didn't report it because it's not obviously an upstream problem, more like a side effect that happens to tickle how auto-session is doing something.
